pytopo/rf/alazar_acquisition.py

```diff
@@ -232,22 +358,135 @@ def post_acquire(self):
+            self.buffers_per_acquisition(),
+            self.samples_per_record(),
+            self.number_of_channels,
+        )).astype(self._datadtype)
 
+    def post_acquire(self):
+        """Demodulate the data and average over period of
+        sample_rate//demod_frq rounded up to nearest integer"""
+        data = super().post_acquire()
+        real = (data * 2 * self.cosarr)[:,:self.demod_samples*self.period,:].reshape(
+            -1, self.demod_samples, self.period, self.number_of_channels).mean(axis=-2)
+        imag = (data * 2 * self.sinarr)[:,:self.demod_samples*self.period,:].reshape(
+            -1, self.demod_samples, self.period, self.number_of_channels).mean(axis=-2)
+        return real + 1j * imag
+
+
+class AvgIQCtl(AvgDemodCtl):
+    """
+    A controller that averages over buffers and subsequently
+    demodulates the averaged data and finally averages over all demodulated
+    samples. The data returned has the format Records x Channels.
+
+    The demodulated samples are averaged over a period of
+    sample_rate//demod_frq rounded up to nearest integer
+    and subsequently averaged over all the periods.
+    """
     def data_shape(self):
-        shp = list(super().data_shape())
+        """
+        Shape of the data that this controller will produce
 
+        Returns:
+            A tuple of the sizes of the data dimensions.
+        """
         return (self.records_per_buffer(),
                 self.number_of_channels)
 
     def data_dims(self):
+        """
+        Dimensions of the data produced
+
+        Returns:
+             A tuple of the names of dimensions of the data returned
+        """
         return ('records', 'channels')
 
     def post_acquire(self):
+        """
+        Average data from super method over all periods.
+        """
         return super().post_acquire().mean(axis=1)
 
 
+class AvgRecIQCtl(AvgRecDemodCtl):
+    """
+    A controller that averages over records and subsequently
+    demodulates the averaged data and finally averages over all demodulated
+    samples. The data returned has the format Buffers x Channels.
+
+    The demodulated samples are averaged over a period of
+    sample_rate//demod_frq rounded up to nearest integer
+    and subsequently averaged over all the periods.
+    """
+    def data_shape(self):
+        """
+        Shape of the data that this controller will produce
+
+        Returns:
+            A tuple of the sizes of the data dimensions.
+        """
+        return (self.buffers_per_acquisition(),
+                self.number_of_channels)
 
+    def data_dims(self):
+        """
+        Dimensions of the data produced
+
+        Returns:
+             A tuple of the names of dimensions of the data returned
+        """
+        return ('buffers', 'channels')
+
+    def post_acquire(self):
+        """
+        Average data from super method over all periods.
+        """
+        return super().post_acquire().mean(axis=1)
```

Comment:
This depends on calling data_shape on the super class to set up the relevant cos/sin arrays. Not sure how that ever worked but it's definitly wrong. This is the same for the matching buffer avg controller AvgIQCtl

Reply:
I pushed a "fix" for this but I think the code should be restructured a bit so don't merge just yet.